### PR TITLE
add test for #23189

### LINKED ISSUE #23189: internal compiler error: 'rustc' panicked at 'No def'n found for DefId { krate: 0, node: 4 } in tcx.tcache'
Just started experimenting with Rust, but as soon as I try to use a Module, compilation fails.

Code:
main.rs:
mod person;

fn main() {
        let p = person{ name : "Example" };
        p.say_hello();
}

person.rs:
struct person {
       name : String,
}

impl person {
        fn say_hello(&self){
                  println!("Hello {}", self.name);
        }
}

---

>  Compiling test v0.0.1 (file:///home/***_/work/test)
>   Running `rustc main.rs --crate-name test --crate-type bin -g --out-dir /home/**__/work/test/target >--emit=dep-info,link -L dependency=/home/_**_/work/test/target -L dependency=/home/**_*/work/test/target/deps`
> error: internal compiler error: unexpected panic
> note: the compiler unexpectedly panicked. this is a bug.
> note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
> note: run with`RUST_BACKTRACE=1` for a backtrace
> thread 'rustc' panicked at 'No def'n found for DefId { krate: 0, node: 4 } in tcx.tcache', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc/middle/ty.rs:5141
> 
> Could not compile `test`.
> 
> Caused by:
>  Process didn't exit successfully: `rustc main.rs --crate-name test --crate-type bin -g --out-dir >/home/****/work/test/target --emit=dep-info,link -L dependency=/home/****/work/test/target -L >dependency=/home/****/work/test/target/deps` (exit code: 101)

---

rustc 1.0.0-nightly (2fc8b1e7c 2015-03-07) (built 2015-03-08)
binary: rustc
commit-hash: 2fc8b1e7c4ca741e59b144c331d69bf189759452
commit-date: 2015-03-07
build-date: 2015-03-08
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly

---

Backtrace:
   Compiling test v0.0.1 (file:///home/****/work/test)
thread 'rustc' panicked at 'No def'n found for DefId { krate: 0, node: 4 } in tcx.tcache', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/librustc/middle/ty.rs:5141

stack backtrace:
   1:     0x7f8c5cdaa2d2 - sys::backtrace::write::h49f84a63c90f3a72QBA
   2:     0x7f8c5cdd0322 - panicking::on_panic::hd99ba6230147c149jHJ
   3:     0x7f8c5cd16a69 - rt::unwind::begin_unwind_inner::hcba984cddf911653mnJ
   4:     0x7f8c5cd16e11 - rt::unwind::begin_unwind_fmt::h056d56048d7d024dXlJ
   5:     0x7f8c5adf820c - middle::ty::lookup_item_type::h6ab48efb47b1c189TU7
   6:     0x7f8c5bbf89a3 - check::check_expr_with_unifier::h7241075278574285947
   7:     0x7f8c5bc1a62d - check::check_decl_local::h554e5d2066e836d5IWr
   8:     0x7f8c5bbd2190 - check::check_block_with_expected::h4d6fd0761db904b4P2r
   9:     0x7f8c5bbba2d2 - check::check_fn::hbe3c2620378bc01ffmn
  10:     0x7f8c5bbcf943 - check::check_bare_fn::h560cf773e203be14Vbn
  11:     0x7f8c5bbc7f1f - check::check_item::ha45b9554a5eca7eazun
  12:     0x7f8c5bc878b3 - check_crate::closure.35693
  13:     0x7f8c5bc833d7 - check_crate::h2872530e4373c1cbhfC
  14:     0x7f8c5d3c585e - driver::phase_3_run_analysis_passes::hd173921852ab87d4mGa
  15:     0x7f8c5d3aa6d5 - driver::compile_input::h552c440872c5b657Nba
  16:     0x7f8c5d46ef1d - run_compiler::h56042a784d7b1fc7G6b
  17:     0x7f8c5d46cc8c - thunk::F.Invoke<A, R>::invoke::h15113937807686641913
  18:     0x7f8c5d46b8e0 - rt::unwind::try::try_fn::h14840337430372503284
  19:     0x7f8c5ce3c0f8 - rust_try_inner
  20:     0x7f8c5ce3c0e5 - rust_try
  21:     0x7f8c5d46c06b - thunk::F.Invoke<A, R>::invoke::h6250943894687305386
  22:     0x7f8c5cdbdb25 - sys::thread::thread_start::h848d60956d3aea28t8E
  23:     0x7f8c56e2a373 - start_thread
  24:     0x7f8c5c99f27c - __clone
  25:                0x0 - <unknown>

### PR DESCRIPTION
**Pull request by euclio**
Fixes #23189

### CODE REVIEW

File: src/test/ui/issues/issue-23189/issue-23189.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+mod auxiliary {
+    mod module;
+}
+
+fn main() {
+    let _ = module { x: 0 }; //~ERROR cannot find struct
```
Comment (tmandry) [inline src/test/ui/issues/issue-23189/issue-23189.rs]: After looking at the original issue, I'm wondering if the ICE only triggered when the name `module` was in scope. Putting it under `auxiliary` would change that.

Did you manage to reproduce the original ICE with this test?

**Comment by tmandry** [comment]
Thanks for your PR!

I did some checking, and I think you need a `use auxiliary::module;` line in your main file to trigger the original ICE.

Note that you can pull the compiler from the original issue with rustup: `rustup toolchain install nightly-2015-03-08`

@bors r? tmandry

**Comment by euclio** [comment]
Thanks for the review. I found that the auxiliary module isn't actually necessary at all, so I pushed a simplified test case.

**Comment by tmandry** [comment]
Nice, even better!

@bors r+ rollup

**Pull request merged**